### PR TITLE
AP_NavEKF3: use onGroundNotMoving for zero velocity fusion

### PR DESCRIPTION
**Pull request by andyp1per**
## Summary

- Fuse synthetic zero velocity when stationary on the ground to constrain gyro bias and Z-axis accel bias learning. XY accel biases remain unobservable on the ground and are inhibited by dvelBiasAxisInhibit
- Applies in all aiding modes: AID_NONE uses zero velocity instead of the large _noaidHorizNoise fake velocity, and AID_RELATIVE/AID_ABSOLUTE modes fuse zero velocity when no real velocity sensor data is available (e.g. optical flow configured but stationary)
- Gated on onGroundNotMoving to avoid injecting zero velocity when the vehicle is being moved
- **Inhibit all accel bias learning when on ground and moving** (e.g. carried on a utility belt or on a boat). Previously only the gravity-alignment check was used, which allowed the gravity-aligned axis to learn biases from external motion. Adding onGroundNotMoving to dvelBiasAxisInhibit prevents incorrect biases that could cause velocity drift after takeoff
- Broaden updateMovementCheck() to run unconditionally when on ground (previously gated on yaw source type)

## Test plan

- [x] EK3_AccelBiasInhibitOnGroundMoving - verifies bias NOT learned during ground movement (SIM_SHIP), and IS learned when stationary. **Fails on master, passes with fix**
- [x] EK3_AccelBiasZeroVelOptFlow - verifies Z-bias convergence with optical flow config (no GPS) via zero velocity fusion
- [x] EK3AccelBias - existing test still passes (no regression)
- [x] Verify no regression in normal GPS flight

### CODE REVIEW

**Comment by rmackay9** [comment]
Great to see the progress on this.  Just one thing, let's be sure this goes through a dev call before being merged.  The AHRS/EKF changes are critical and complex

File: libraries/AP_NavEKF3/AP_NavEKF3_core.cpp [hunk 1]
```diff
@@ -1165,8 +1166,14 @@ void NavEKF3_core::CovariancePrediction(Vector3F *rotVarVecPtr)
...
-            // Don't attempt learning of IMU delta velocty bias if on ground and not aligned with the gravity vector
-            const bool is_bias_observable = (fabsF(prevTnb[index][2]) > 0.8f) || !onGround;
+            // Don't attempt learning of IMU delta velocity bias if on ground.
+            // In flight: all axes are observable from velocity/position aiding.
+            // On ground and stationary: only the gravity-aligned axis (Z for a level
+            // vehicle) is observable. XY biases remain unobservable until the vehicle
+            // accelerates horizontally in flight.
+            // On ground and moving (e.g. carried or on a boat): inhibit all axes
+            // to prevent learning biases from external motion accelerations.
+            const bool is_bias_observable = (fabsF(prevTnb[index][2]) > 0.8f && onGroundNotMoving) || !onGround;
```
Comment (IamPete1) [inline libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:1176]: Surely you cannot be `onGroundNotMoving` without also being `onGround`. Can we remove `|| !onGround`?

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 2]
```diff
@@ -672,13 +674,63 @@ void NavEKF3_core::SelectVelPosFusion()
...
+            // Use onGroundNotMoving to avoid fusing zero velocity when the vehicle
+            // is being moved (e.g. on a boat or carried by hand).
+            // takeoff_expected covers the armed-on-ground case before liftoff.
+            const bool onGroundNotFlying = onGroundNotMoving || dal.get_takeoff_expected();
+            if (onGroundNotFlying && tiltAlignComplete) {
+                fuseVelData = true;
+                fusingStationaryZeroVel = true;
+                velPosObs[0] = 0.0f;
+                velPosObs[1] = 0.0f;
+                velPosObs[2] = 0.0f;
```
Comment (IamPete1) [inline libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:690]: Why fuse zero vel rather than not fusing vel at all?

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 3]
```diff
@@ -672,13 +674,63 @@ void NavEKF3_core::SelectVelPosFusion()
...
+    // learning. XY accel biases are unobservable on the ground and are inhibited by
+    // dvelBiasAxisInhibit. Without this, configurations like optical flow where
+    // PV_AidingMode is AID_RELATIVE but no velocity data is available when stationary
+    // have no velocity observations at all, causing unchecked bias drift. The timeout
+    // check on each velocity source ensures we only inject zero velocity when no real
+    // sensor data is being fused. Use onGroundNotMoving to avoid injecting zero velocity
+    // when the vehicle is being moved, and takeoff_expected for armed-on-ground.
+    // Gate behind fuseHgtData to limit fusion rate to baro rate (~10Hz) and avoid
+    // overconstraining the filter by fusing at IMU rate.
+    const bool onGroundNotFlying2 = onGroundNotMoving || dal.get_takeoff_expected();
```
Comment (IamPete1) [inline libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: Why not `onGroundNotFlying` what is the `2` for?

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp [hunk 4]
```diff
@@ -713,7 +765,14 @@ void NavEKF3_core::FuseVelPosNED()
         // estimate the velocity, horiz position and height measurement variances.
         // Use different errors if operating without external aiding using an assumed position or velocity of zero
         if (PV_AidingMode == AID_NONE) {
-            if (tiltAlignComplete && motorsArmed) {
+            if (fusingStationaryZeroVel) {
+                // Synthetic zero velocity on ground: use 1 m/s noise (variance = 1 m^2/s^2).
+                // This is tighter than _noaidHorizNoise (default 10 m/s) used when armed,
```
Comment (IamPete1) [inline libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:770]: The param range for `NOAID_M_NSE` is 0.5 to 10. So should this be min of that and 1?

File: Tools/autotest/arducopter.py [hunk 5]
```diff
@@ -1714,6 +1714,142 @@ def EK3AccelBias(self):
...
+    def EK3_AccelBiasInhibitOnGroundMoving(self):
+        '''Test EKF3 inhibits accel bias learning when vehicle is moved on ground'''
+        # When a copter is on the ground and being moved (carried, on a ship),
+        # the EKF should not learn accel biases from the external motion.
+        # Without the fix (onGroundNotMoving check in dvelBiasAxisInhibit),
+        # the Z-axis bias is learnable on the ground (passes gravity alignment
+        # check), and GPS velocity innovations can drive incorrect bias learning
+        # during ground movement. With the fix, all bias learning is inhibited
+        # when onGround && !onGroundNotMoving.
+        self.context_push()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
```

File: Tools/autotest/arducopter.py [hunk 6]
```diff
@@ -1714,6 +1714,142 @@ def EK3AccelBias(self):
...
+        self.delay_sim_time(60)
+        self.assert_dataflash_message_field_level_at(
+            "XKF2", "AZ", 0.7,
+            condition="XKF2.C==1",
+            maintain=1,
+            tolerance=0.3,
+        )
+
+        self.context_pop()
+        self.reboot_sitl()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by rmackay9** [comment]
@andyp1per, could you explain what value this adds?

**Comment by tridge** [comment]
@andyp1per can you explain what situation this is trying to address?

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:690]: because we know that velocity is 0 which is not the same as not fusing velocity at all. Fusing 0 velocity is what stops the biases running away while sitting on the ground.

**Comment by andyp1per** [comment]
@tridge it's required by https://github.com/ArduPilot/ardupilot/pull/32471 but I am trying to split the PR's up into manageable chunks.

**Comment by priseborough** [comment]
The underlying assumption of the EKF when not using GPS ad on ground is that the vehicle is relatively static, ie OK for operator handling, but not vigorous or continuous movement. This PR improves the reliability of the accel bias learning in this environment.

Log showing before and after would be good. The benefit of this could be demonstrated using replay which has the benefit of removing test variability.

Autotest shows the effect.

**Comment by tridge** [comment]
@priseborough I didn't see why the changes only applied in a non-GPS (and presumably non-flow?) setup, I've probably just missed a condition in the code that makes this not impact GPS (or other velocity sensor) enabled setups

File: libraries/AP_NavEKF3/AP_NavEKF3_core.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:1176]: we can observe biases inflight or if we are on the ground and not moving - we need both checks. (it's !onGround rather than onGround)

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp]: Fixed thanks

File: libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp (hunk 4 above)
Comment (andyp1per) [reply libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:770]: Yes, nice catch - thanks, fixed.

File: Tools/autotest/arducopter.py (hunk 5 above)
Comment (andyp1per) [reply Tools/autotest/arducopter.py]: Thanks, fixed

File: Tools/autotest/arducopter.py (hunk 6 above)
Comment (andyp1per) [reply Tools/autotest/arducopter.py]: Thanks, fixed

**Comment by andyp1per** [comment]
@tridge I added a new test: EK3_ZeroVelFusionNotUsedWithGPS

The zero velocity fusion changes do NOT affect GPS-enabled setups because of two independent protection layers:

 1. onGroundNotMoving gate — On a moving platform where IMU detects motion, onGroundNotFlying is false, so the zero velocity code path is never entered.
 2. GPS velocity timeout check — Even when onGroundNotMoving is TRUE (smooth motion below detection thresholds), zero  velocity fusion requires !haveRecentGpsVel, which is false when GPS is providing velocity data. GPS updates  lastVelPassTime_ms every fusion cycle, so the 1-second timeout never triggers.

Test scenario: Vehicle on a fast boat (15 m/s, 1000m diameter circle). The large radius keeps centripetal acceleration (0.45 m/s²) and angular rate (1.7 deg/s) below the movement detection thresholds, so onGroundNotMoving stays TRUE at steady
state. This specifically exercises the code path on line 712 of AP_NavEKF3_PosVelFusion.cpp where the GPS timeout check
  prevents zero velocity injection.

Assertions:
 - EKF groundspeed tracks GPS at ~15 m/s (not pulled to zero)
 - Z-axis accel bias converges to 0.7 m/s² via GPS velocity observations
 - Velocity remains correct after bias convergence

**Comment by rmackay9** [comment]
I guess asking why we have special checks for GPS vs other position sources (e.g. external nav) is out-of-scope?

**Comment by andyp1per** [comment]
> I guess asking why we have special checks for GPS vs other position sources (e.g. external nav) is out-of-scope?

I'm prepared to believe there are corner cases missed by the existing code as I keep tripping over them! One of the issues is that the behaviour is not usually fatal just unexplained. I wouldn't want to change those corner cases without hardware to test on. You would need a velocity source other than GPS which I think means beacons or some external nav systems

**Review by priseborough — APPROVED** [review]

**Comment by peterbarker** [comment]
@IamPete1 could you make sure you're satisfied with Andy's responses, please?

We're planning on merging after you are.

**Comment by IamPete1** [comment]
> @IamPete1 could you make sure you're satisfied with Andy's responses, please?
> 
> We're planning on merging after you are.

Yes, I'm happy.

**Comment by rmackay9** [comment]
I think this has added an unused, "usingVertVel" method.  It does not harm but I'm surprised this got through the review without anyone noticing.
